providers.tf

```diff
@@ -1,3 +1,4 @@
 provider "aws" {
-  # Configuration options
+  source = "hashicorp/aws"
+  region = var.aws_provider_region
```

Comment:
I don't think adding the region is necessary... this style of adding the source is for actual TF config, not modules... and I think adding the region as a variable is likely to introduce some edge cases that don't work quite right...

From the Terraform docs, they show provider declarations inside modules looking more like this:
```
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = ">= 2.7.0"
    }
  }
}
```

Reply:
@TyBrown Even with region locally specified in the main.tf:
```
terraform {
  backend "s3" {
    bucket         = "replicated-staging-cloud-custodian-terraform-state"
    key            = "cloud-custodian-terraform/terraform.tfstate"
    region         = "us-east-1"
    dynamodb_table = "cloud-custodian-iac-state-lock"
    encrypt        = true
  }
  required_version   = ">= 1.0.11"
  required_providers {
    aws = {
      source = "hashicorp/aws"
    }
  }
}

provider "aws" {
  region = "us-east-1"
}

module "aws_event_handler_notify_unencrypted_ebs" {
  source = "github.com/replicatedhq/aws-event-handler-terraform-module"

  handler_entrypoint = "print_decoded"
  handler_filepath   = "../../../scripts/aws/notify-ebs-unencrypted-volumes.py"
  handler_runtime    = "python3.9"
  name               = "cloud-custodian-notify-ebs-unencrypted-volumes"
  owner              = "sre"
}
```
I still get the error relating to the provider block of the source module:
```
│ Error: error configuring Terraform AWS Provider: error validating provider credentials: error calling sts:GetCallerIdentity: operation error STS: GetCallerIdentity, failed to resolve service endpoint, an AWS region is required, but was not found
│
│   with module.aws_event_handler_notify_unencrypted_ebs.provider["registry.terraform.io/hashicorp/aws"],
│   on .terraform/modules/aws_event_handler_notify_unencrypted_ebs/providers.tf line 1, in provider "aws":
│    1: provider "aws" {
│
```

Reply:
I also still get the warning about not specifying source with it defined in the main.tf of the consuming module

Reply:
removed the source key b/c adding region got past the warning, but I still think region is necessary to get past the error I mentioned

Reply:
I suspect that because the `source` in your example there is still pointing at GitHub, that TF is pulling the `main` branch from GitHub... in order to test out the changes I proposed above, I think you'll need to set the module source in your testing TF to be a local path so that TF actually uses the local checkout of the module that you're working on.

Reply:
@TyBrown I tested pointing to my local branch where I've made these changes before my last push, and I get a successful plan: 
```
> cat main.tf
terraform {
  backend "s3" {
    bucket         = "replicated-staging-cloud-custodian-terraform-state"
    key            = "cloud-custodian-terraform/terraform.tfstate"
    region         = "us-east-1"
    dynamodb_table = "cloud-custodian-iac-state-lock"
    encrypt        = true
  }
  required_version   = ">= 1.0.11"
  required_providers {
    aws = {
      source = "hashicorp/aws"
    }
  }
}

module "aws_event_handler_notify_unencrypted_ebs" {
  source = "/Users/chasehippen/git/aws-event-handler-terraform-module/"
  # source = "github.com/replicatedhq/aws-event-handler-terraform-module"

  handler_entrypoint  = "print_decoded"
  handler_filepath    = "../../../scripts/aws/notify-ebs-unencrypted-volumes.py"
  handler_runtime     = "python3.9"
  name                = "check_ebs"
  owner               = "sre"
  aws_provider_region = "us-east-1"
}
```
```
> cat /Users/chasehippen/git/aws-event-handler-terraform-module/providers.tf
provider "aws" {
  region = var.aws_provider_region
}
```

Reply:
Re-reading this initial warning:
```
 Earlier versions of Terraform used empty provider blocks ("proxy provider configurations") for child modules to declare their need to be passed a provider configuration by their
 callers. That approach was ambiguous and is now deprecated.

 If you control this module, you can migrate to the new declaration syntax by removing all of the empty provider "aws" blocks and then adding or updating an entry like the
 following to the required_providers block of module.aws_event_handler_notify_unencrypted_ebs:
 aws = {
 source = "hashicorp/aws"
 }
```
I think we should just delete the `provider "aws" {}` block entirely from this module - that'll get rid of the warning, and doesn't cause the region error either (the module will use the implicit provider that's being used in the calling TF).

To ensure that we declare the module as being necessary, we can replace the `provider "aws" {}` block with a Terraform block that declares the minimum required provider version:
```
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = ">= 2.7.0"
    }
  }
}
```

In addition, we should add the `archive` and `null` providers to this `required_providers` block as well, since I see those in the README as being necessary.

Reply:
@TyBrown If I take out the provider block entirely I get an error about missing a provider block in addition to the region error:
```
╷
│ Error: Invalid provider configuration
│
│ Provider "registry.terraform.io/hashicorp/aws" requires explicit configuration. Add a provider block to the root module and configure the provider's required arguments as
│ described in the provider documentation.
│
╵
╷
│ Error: error configuring Terraform AWS Provider: error validating provider credentials: error calling sts:GetCallerIdentity: operation error STS: GetCallerIdentity, failed to resolve service endpoint, an AWS region is required, but was not found
│
│   with provider["registry.terraform.io/hashicorp/aws"],
│   on <empty> line 0:
│   (source code not available)
│
```